### ISSUE TITLE
NESTJS - `Error: 12 UNIMPLEMENTED: The server does not implement this method`

### ISSUE THREAD

**Issue by jessequinn** [stephenh/ts-proto#63]
I am using the new nestjs proto and i implemented the interfaces on the server and client however i am getting `Error: 12 UNIMPLEMENTED: The server does not implement this method`.

The following code is for client, server, and my proto file in TS.

```
import { Inject, Injectable, Logger, OnModuleInit } from '@nestjs/common';
import { ClientGrpc } from '@nestjs/microservices';
import { Observable } from 'rxjs';
import {
  US_CITIES_PACKAGE_NAME,
  US_CITIES_SERVICE_PROTO_SERVICE_NAME,
  USCitiesResponse,
  USCitiesServiceProtoClient,
  USCityResponse,
} from '../interfaces/us-cities';

@Injectable()
export class USCitiesService implements OnModuleInit {
  private readonly logger = new Logger(USCitiesService.name);
  private uSCitiesServiceProtoClient: USCitiesServiceProtoClient;

  constructor(@Inject(US_CITIES_PACKAGE_NAME) private client: ClientGrpc) {}

  onModuleInit() {
    this.uSCitiesServiceProtoClient = this.client.getService<
      USCitiesServiceProtoClient
    >(US_CITIES_SERVICE_PROTO_SERVICE_NAME);
  }

  getUSCity(id: number): Observable<USCityResponse> {
    return this.uSCitiesServiceProtoClient.getUSCity({ id });
  }

  getUSCities(): Observable<USCitiesResponse> {
    return this.uSCitiesServiceProtoClient.getUSCities({});
  }
}
```

```
import { Controller, Logger } from '@nestjs/common';
import { USCitiesService } from './us-cities.service';
import {
  USCitiesResponse, USCitiesServiceProtoController, USCitiesServiceProtoControllerMethods,
  USCityRequest,
  USCityResponse,
} from '../interfaces/us-cities';

@Controller()
@USCitiesServiceProtoControllerMethods()
export class USCitiesController implements USCitiesServiceProtoController {
  private readonly logger = new Logger(USCitiesController.name);
  constructor(private uSCitiesService: USCitiesService) {}

  async getUSCity(data: USCityRequest): Promise<USCityResponse> {
    this.logger.debug('accessing GetUSCity');
    const result = await this.uSCitiesService.getUSCity(data.id);
    return { uSCity: result };
  }

  async getUSCities(): Promise<USCitiesResponse> {
    this.logger.debug('accessing GetUSCities');
    const results = await this.uSCitiesService.getUSCities();
    return {
      uSCities: results,
    };
  }
}
```

```
/* eslint-disable */
import { Observable } from 'rxjs';
import { Empty } from './google/protobuf/empty';
import { GrpcMethod, GrpcStreamMethod } from '@nestjs/microservices';

export interface USCityRequest {
  id: number;
}

export interface USCitiesResponse {
  uSCities: USCity[];
}

export interface USCityResponse {
  uSCity: USCity | undefined;
}

export interface USCity {
  id: number;
  idState: number;
  city: string;
  county: string;
  latitude: number;
  longitude: number;
}

export interface USCitiesServiceProtoController {
  getUSCity(
    request: USCityRequest,
  ): Promise<USCityResponse> | Observable<USCityResponse> | USCityResponse;

  getUSCities(
    request: Empty,
  ):
    | Promise<USCitiesResponse>
    | Observable<USCitiesResponse>
    | USCitiesResponse;
}

export interface USCitiesServiceProtoClient {
  getUSCity(request: USCityRequest): Observable<USCityResponse>;

  getUSCities(request: Empty): Observable<USCitiesResponse>;
}

export function USCitiesServiceProtoControllerMethods() {
  return function(constructor: Function) {
    const grpcMethods: string[] = ['getUSCity', 'getUSCities'];
    for (const method of grpcMethods) {
      const descriptor: any = Reflect.getOwnPropertyDescriptor(
        constructor.prototype,
        method,
      );
      GrpcMethod('USCitiesServiceProto', method)(
        constructor.prototype[method],
        method,
        descriptor,
      );
    }
    const grpcStreamMethods: string[] = [];
    for (const method of grpcStreamMethods) {
      const descriptor: any = Reflect.getOwnPropertyDescriptor(
        constructor.prototype,
        method,
      );
      GrpcStreamMethod('USCitiesServiceProto', method)(
        constructor.prototype[method],
        method,
        descriptor,
      );
    }
  };
}

export const US_CITIES_PACKAGE_NAME = 'us.cities';
export const US_CITIES_SERVICE_PROTO_SERVICE_NAME = 'USCitiesServiceProto';
```

**Comment by stephenh** [comment]
@iangregsondev @ToonvanStrijp any ideas re ^? In general @jessequinn more of the stack trace than just the one-line message would be helpful.

**Comment by jessequinn** [comment]
the problem is the decorator
// @USStatesServiceProtoControllerMethods()

**Comment by iangregsondev** [comment]
Hi.

So the issue is the decorator ? This is your decorator ?

All we are doing is creating interfaces so it doesn’t look like anything we can help with.

Let me know if I am not following and I will try and dig into it later.

Cheers.

**Comment by ToonvanStrijp** [comment]
@jessequinn The handler `GetUSState` is not getting generated in your ts files?

Can you show me your `.proto` file? Did you maybe forgot to `regenerate` the `typescript` files?

**Comment by jessequinn** [comment]
```
syntax = "proto3";

package us.cities;

import "google/protobuf/empty.proto";

service USCitiesServiceProto {
    rpc GetUSCity(USCityRequest) returns (USCityResponse) {}
    rpc GetUSCities(google.protobuf.Empty) returns (USCitiesResponse) {}
}

message USCityRequest {
    int32 id = 1;
}

message USCitiesResponse {
    repeated USCity uSCities = 1;
}

message USCityResponse {
    USCity uSCity = 1;
}

message USCity {
    int32 id = 1;
    int32 idState = 2;
    string city = 3;
    string county = 4;
    double latitude = 5;
    double longitude = 6;
}
```

```
syntax = "proto3";

package us.states;

import "google/protobuf/empty.proto";

service USStatesServiceProto {
    rpc GetUSState (USStateRequest) returns (USStateResponse) {}
    rpc GetUSStates(google.protobuf.Empty) returns (USStatesResponse) {}
}

message USStateRequest {
    int32 id = 1;
}

message USStatesResponse {
    repeated USState uSStates = 1;
}

message USStateResponse {
    USState uSState = 1;
}

message USState {
    int32 id = 1;
    string stateCode = 2;
    string stateName = 3;
}
```

**Comment by ToonvanStrijp** [comment]
@jessequinn I really need some more information or see more configuration files of you you initialize the modules etc.

From what I can see now, you have two proto files. Did you generate them both? I think it's an configuration issue, and that your setup isn't generating all the files it should.

**Comment by jessequinn** [comment]
i run the following command

```
protoc \
--plugin=./node_modules/.bin/protoc-gen-ts_proto \
--ts_proto_out=./build *.proto \
--ts_proto_opt=outputEncodeMethods=false,outputJsonMethods=false,outputClientImpl=false \
--ts_proto_opt=nestJs=true
```

to generate the TS files.
what other information do you need?

**Comment by jessequinn** [comment]
these are the outputted TS files

```/* eslint-disable */
import { Observable } from 'rxjs';
import { Empty } from './google/protobuf/empty';
import { GrpcMethod, GrpcStreamMethod } from '@nestjs/microservices';


export interface USCityRequest {
  id: number;
}

export interface USCitiesResponse {
  uSCities: USCity[];
}

export interface USCityResponse {
  uSCity: USCity | undefined;
}

export interface USCity {
  id: number;
  uSState: USState | undefined;
  city: string;
  county: string;
  latitude: number;
  longitude: number;
}

export interface USState {
  id: number;
  stateCode: string;
  stateName: string;
}

export interface USCitiesServiceProtoController {

  getUSCity(request: USCityRequest): Promise<USCityResponse> | Observable<USCityResponse> | USCityResponse;

  getUSCities(request: Empty): Promise<USCitiesResponse> | Observable<USCitiesResponse> | USCitiesResponse;

}

export interface USCitiesServiceProtoClient {

  getUSCity(request: USCityRequest): Observable<USCityResponse>;

  getUSCities(request: Empty): Observable<USCitiesResponse>;

}

export function USCitiesServiceProtoControllerMethods() {
  return function (constructor: Function) {
    const grpcMethods: string[] = ['getUSCity', 'getUSCities'];
    for (const method of grpcMethods) {
      const descriptor: any = Reflect.getOwnPropertyDescriptor(constructor.prototype, method);
      GrpcMethod('USCitiesServiceProto', method)(constructor.prototype[method], method, descriptor);
    }
    const grpcStreamMethods: string[] = [];
    for (const method of grpcStreamMethods) {
      const descriptor: any = Reflect.getOwnPropertyDescriptor(constructor.prototype, method);
      GrpcStreamMethod('USCitiesServiceProto', method)(constructor.prototype[method], method, descriptor);
    }
  }
}

export const US_CITIES_PACKAGE_NAME = 'us.cities'
export const U_SCITIES_SERVICE_PROTO_SERVICE_NAME = 'USCitiesServiceProto';
```

```
/* eslint-disable */
import { Observable } from 'rxjs';
import { Empty } from './google/protobuf/empty';
import { GrpcMethod, GrpcStreamMethod } from '@nestjs/microservices';


export interface USStateRequest {
  id: number;
}

export interface USStatesResponse {
  uSStates: USState[];
}

export interface USStateResponse {
  uSState: USState | undefined;
}

export interface USState {
  id: number;
  stateCode: string;
  stateName: string;
  cities: USCity[];
}

export interface USCity {
  id: number;
  uSState: USState | undefined;
  city: string;
  county: string;
  latitude: number;
  longitude: number;
}

export interface USStatesServiceProtoController {

  getUSState(request: USStateRequest): Promise<USStateResponse> | Observable<USStateResponse> | USStateResponse;

  getUSStates(request: Empty): Promise<USStatesResponse> | Observable<USStatesResponse> | USStatesResponse;

}

export interface USStatesServiceProtoClient {

  getUSState(request: USStateRequest): Observable<USStateResponse>;

  getUSStates(request: Empty): Observable<USStatesResponse>;

}

export function USStatesServiceProtoControllerMethods() {
  return function (constructor: Function) {
    const grpcMethods: string[] = ['getUSState', 'getUSStates'];
    for (const method of grpcMethods) {
      const descriptor: any = Reflect.getOwnPropertyDescriptor(constructor.prototype, method);
      GrpcMethod('USStatesServiceProto', method)(constructor.prototype[method], method, descriptor);
    }
    const grpcStreamMethods: string[] = [];
    for (const method of grpcStreamMethods) {
      const descriptor: any = Reflect.getOwnPropertyDescriptor(constructor.prototype, method);
      GrpcStreamMethod('USStatesServiceProto', method)(constructor.prototype[method], method, descriptor);
    }
  }
}

export const US_STATES_PACKAGE_NAME = 'us.states'
export const U_SSTATES_SERVICE_PROTO_SERVICE_NAME = 'USStatesServiceProto';
```

**Comment by ToonvanStrijp** [comment]
@jessequinn I would be nice to have a sample repo with this issue in it so we can reproduce it. Otherwise we have to build it ourselves and that will take more time. So I you can send a zip with the project containing the error that we can run and debug would be nice.

**Comment by jessequinn** [comment]
ill try to get it this to you this weekend.

**Comment by stephenh** [comment]
Does it matter that these error lines you're getting have upper-case `Get`:

```
Method handler GetUSState for /us.states.USStatesServiceProto/GetUSState expected but not provided
Method handler GetUSStates for /us.states.USStatesServiceProto/GetUSStates expected but not provided
```

But we're generating lower-case `get` methods & decorators?

**Comment by stephenh** [comment]
Maybe we should upper-case the `method` that we pass here:

```
    const grpcMethods: string[] = ['getUSState', 'getUSStates'];
    for (const method of grpcMethods) {
      const descriptor: any = Reflect.getOwnPropertyDescriptor(constructor.prototype, method);
      GrpcMethod('USStatesServiceProto', method)(constructor.prototype[method], method, descriptor);
    }
```

But not sure why this would be tripping @jessequinn up but not @ToonvanStrijp who I assume is using the generated-decorator successfully.

**Comment by ToonvanStrijp** [comment]
@stephenh good catch! They indeed should start with a capital letter indeed! I'll create a fix for this later today ;)

**Comment by scidoc-coder** [comment]
@ToonvanStrijp please verify this as i believe i found the problem. The issue occurs when your service has functions that have more than one consecutive capital letter.

```
service USStatesServiceProto {
    rpc Check1 (USStateRequest) returns (USStateResponse) {}
    rpc GetUSStates (google.protobuf.Empty) returns (USStatesResponse) {}
}
```

`Check1` works fine with the `USStatesServiceProtoControllerMethods` as a decorator; however `GetUSStates` as it is looking for `getUsStates` and not `getUSStates`

**Comment by lukas-andre** [comment]
I have the same problem and looking for the reason why I found this https://github.com/nestjs/nest/issues/2659 , when there is a server-side streming the decorator must be used @GrpcMethod instead of @GrpcStreamMethod that could fix this issue

**Comment by lukas-andre** [comment]
I'm not sure if that's the case, but I had a similar problem and I conclude that it is due to the generated driver decorator, when they transform a server side rpc stream they are making the method use the ``` @GrpcStreamMethod()``` decorator, and According to this https://github.com/nestjs/nest/issues/2659 , the decorator to use is `` @GrpcMethod()``, I set the ```@GrpcMethod()``` in all methods that return an Observable or Stream and the service recognized them without problem. 

Keep in mind that it is if the method receives a stream and sends a stream, then you should use the ```@GrpcStreamMethod``` decorator but I am not entirely sure of it since I have not done tests

**Comment by ssilve1989** [comment]
@lukas-andre Briefly skimmed over this issue and seems related to a PR i've opened to address this. https://github.com/stephenh/ts-proto/pull/68

My PR has tested the cases of rpc calls with:
 - request -> stream
 - stream -> stream
 - stream -> request
 - request -> request

**Comment by stephenh** [comment]
Great, thanks @ssilve1989 . I've released 1.21.1 with your fix. I'm going to optimistically close this as fixed. @jessequinn / @lukas-andre let us know if this didn't fix it for you.
